intent/src/lib.rs

```diff
@@ -132,20 +181,86 @@ impl IntentContractImpl {
         }
     }
 
-    /// Callback which changes a status of the intent.
+    /// Callback which changes a status of the intent. The `amount` arg makes sense in case of
+    /// calling this transaction by `ft_transfer_call`. In case if the intent is expired the amount
+    /// should be the same as the solver send to return the funds back. In successful execution the
+    /// amount should be 0 to prevent refund funds in the `ft_resolve_transfer` callback to the solver.
     ///
     /// # Panics
     ///
     /// Panics if intent with given ID doesn't exist.
     #[private]
-    pub fn change_intent_status(&mut self, intent_id: &String, status: Status) {
+    pub fn change_intent_status(
+        &mut self,
+        intent_id: &String,
+        status: Status,
+        amount: U128,
+    ) -> U128 {
+        log!(
+            "Changing status of the intent with id: {} to {status:?} status",
+            intent_id
+        );
         let intent_with_status = self
             .intents
             .get_mut(intent_id)
             .ok_or_else(|| IntentError::NotFound(intent_id.clone()))
             .unwrap();
 
         intent_with_status.set_status(status);
+        amount
+    }
+
+    /// Callback which finishes creating an intent.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the storage deposit is too low or there is no storage deposit for the initiator.
+    #[private]
+    pub fn finish_creating_intent(
+        &mut self,
+        id: String,
+        amount: U128,
+        intent: Intent,
+        #[callback_result] result: Result<Option<StorageBalance>, near_sdk::PromiseError>,
+    ) -> U128 {
+        match result {
+            Ok(Some(balance)) => {
+                assert!(
+                    balance.total >= MIN_STORAGE_DEPOSIT,
```

Comment:
This check doesn't really provide any guarantees that future transfer to the user would succeed, since this constant `MIN_STORAGE_DEPOSIT` can differ between different fungible token implementations. What if this contract is optimized in some way and requires less NEAR to be deposited for storage? This check would fail and user would not be able to create an intent? Or the opposite: what if it requires more NEAR to be deposited? This check would succeed but execution of this intent would fail.
There is a more generic question: are you sure that all fungible tokens implement Storage Management (NEP-145) standard? What if creators of a ft token decided to pay for storage for their users and implement some different incentive mechanism to pay for storage in this exact token, let's say? I don't know if anyone really does it, just thinking about potential consequences

Reply:
This intent covers NEP-141 tokens, which have to implement NEP-145 as well. What you mentioned is out of the scope of this intent.